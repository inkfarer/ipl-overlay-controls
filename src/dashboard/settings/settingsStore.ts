import { NodeCGBrowser } from 'nodecg/browser';
import { LastFmSettings, RadiaSettings, RuntimeConfig } from 'schemas';
import { GameVersion } from 'types/enums/gameVersion';
import { SetGameVersionResponse } from 'types/messages/runtimeConfig';
import { defineStore } from 'pinia';

const lastFmSettings = nodecg.Replicant<LastFmSettings>('lastFmSettings');
const radiaSettings = nodecg.Replicant<RadiaSettings>('radiaSettings');
const runtimeConfig = nodecg.Replicant<RuntimeConfig>('runtimeConfig');

export const settingsReps = [lastFmSettings, radiaSettings, runtimeConfig];

export interface SettingsStore {
    lastFmSettings: LastFmSettings
    radiaSettings: RadiaSettings
    runtimeConfig: RuntimeConfig
}

export const useSettingsStore = defineStore('settings', {
    state: () => ({
        lastFmSettings: {},
        radiaSettings: {
            guildID: null,
            enabled: null,
            updateOnImport: null
        },
        runtimeConfig: null
    } as SettingsStore),
    actions: {
        setLastFmSettings({ newValue }: { newValue: LastFmSettings }): void {
            lastFmSettings.value = newValue;
        },
        setRadiaSettings({ newValue }: { newValue: RadiaSettings }): void {
            radiaSettings.value = newValue;
        },
        setUpdateOnImport(updateOnImport: boolean): void {
            radiaSettings.value.updateOnImport = updateOnImport;
        },
        async attemptRadiaConnection(): Promise<void> {
            return nodecg.sendMessage('retryRadiaAvailabilityCheck');
        },
        setGameVersion(newValue: GameVersion): Promise<SetGameVersionResponse> {
            return nodecg.sendMessage('setGameVersion', { version: newValue });
        }
    }
});
