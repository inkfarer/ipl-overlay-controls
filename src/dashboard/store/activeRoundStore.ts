import { ActiveRound, SwapColorsInternally } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { GameWinner } from 'types/enums/gameWinner';
import { SetActiveColorRequest, SetWinnerRequest, SwapRoundColorRequest } from 'types/messages/activeRound';
import { SetRoundRequest } from 'types/messages/rounds';
import { ActiveRoundGame } from 'types/activeRoundGame';

const activeRound = nodecg.Replicant<ActiveRound>('activeRound');
const swapColorsInternally = nodecg.Replicant<SwapColorsInternally>('swapColorsInternally');

export const activeRoundReps = [ activeRound, swapColorsInternally ];

export interface ActiveRoundStore {
    activeRound: ActiveRound;
    swapColorsInternally: SwapColorsInternally;
}

export const activeRoundStore = createStore<ActiveRoundStore>({
    state: {
        activeRound: null,
        swapColorsInternally: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        setWinner(store, { winner }: { winner: GameWinner }): void {
            nodecg.sendMessage('setWinner', { winner });
        },
        removeWinner(): void {
            nodecg.sendMessage('removeWinner');
        },
        setActiveColor(store, request: SetActiveColorRequest): void {
            nodecg.sendMessage('setActiveColor', request);
        },
        swapColors(): void {
            swapColorsInternally.value = !swapColorsInternally.value;
        },
        setWinnerForIndex(store, { index, winner }: { index: number, winner: GameWinner }): void {
            nodecg.sendMessage('setWinner', { winner, roundIndex: index } as SetWinnerRequest);
        },
        resetActiveRound(): void {
            nodecg.sendMessage('resetActiveRound');
        },
        setActiveRound(store, request: SetRoundRequest): void {
            nodecg.sendMessage('setActiveRound', request);
        },
        swapRoundColor(store, request: SwapRoundColorRequest): void {
            nodecg.sendMessage('swapRoundColor', request);
        },
        updateActiveGames(store, games: ActiveRoundGame[]): void {
            nodecg.sendMessage('updateActiveGames', { games });
        },
        switchToNextColor(): void {
            nodecg.sendMessage('switchToNextColor');
        },
        switchToPreviousColor(): void {
            nodecg.sendMessage('switchToPreviousColor');
        },
        async getNextAndPreviousColors(): Promise<void> {
            return nodecg.sendMessage('getNextAndPreviousColors');
        }
    }
});

export const activeRoundStoreKey: InjectionKey<Store<ActiveRoundStore>> = Symbol();

export function useActiveRoundStore(): Store<ActiveRoundStore> {
    return useStore(activeRoundStoreKey);
}
