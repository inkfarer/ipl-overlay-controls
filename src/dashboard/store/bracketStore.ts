import { Bracket } from '@tourneyview/common';
import { defineStore } from 'pinia';

const bracketData = nodecg.Replicant<Bracket | undefined>('bracketData');

export const bracketReps = [ bracketData ];

export interface BracketStore {
    bracketData: Bracket | undefined;
}

export const useBracketStore = defineStore('brackets', {
    state: () => ({
        bracketData: undefined
    } as BracketStore),
    actions: {
        resetBracketData() {
            bracketData.value = undefined;
        }
    }
});
