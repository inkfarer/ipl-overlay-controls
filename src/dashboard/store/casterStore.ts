import { NodeCGBrowser } from 'nodecg/browser';
import { Caster, Casters, RadiaSettings } from 'schemas';
import { generateId } from '../../helpers/generateId';
import { defineStore } from 'pinia';

const casters = nodecg.Replicant<Casters>('casters');
const radiaSettings = nodecg.Replicant<RadiaSettings>('radiaSettings');

export const castersReps = [ casters, radiaSettings ];

export interface CasterStore {
    casters: Casters;
    uncommittedCasters: Casters;
    radiaSettings: RadiaSettings;
}

const defaultCaster = {
    name: 'New Caster',
    twitter: '',
    pronouns: '?/?'
};

export const useCasterStore = defineStore('casters', {
    state: () => ({
        casters: {},
        uncommittedCasters: {},
        radiaSettings: {
            enabled: null,
            guildID: null,
            updateOnImport: null
        }
    } as CasterStore),
    actions: {
        updateCaster({ id, newValue }: { id: string, newValue: Caster }): void {
            casters.value[id] = newValue;
        },
        removeUncommittedCaster(id: string): void {
            delete this.uncommittedCasters[id];
        },
        addUncommittedCaster({ id, caster }: { id: string, caster?: Caster }): void {
            this.uncommittedCasters[id] = caster ?? defaultCaster;
        },
        addDefaultCaster(): string {
            const id = generateId();
            this.addUncommittedCaster({ id });
            return id;
        },
        async saveUncommittedCaster({ id, caster }: { id: string, caster: Caster }): Promise<string> {
            const newId = await nodecg.sendMessage('saveCaster', caster);
            this.removeUncommittedCaster(id);
            return newId;
        },
        async loadCastersFromVc(): Promise<void> {
            const result = await nodecg.sendMessage('getLiveCommentators');
            if (result.extra && result.extra.length > 0) {
                for (let i = 0; i < result.extra.length; i++) {
                    const extraCaster = result.extra[i];
                    const id = extraCaster.discord_user_id;
                    delete extraCaster.discord_user_id;
                    this.addUncommittedCaster({ id, caster: extraCaster });
                }
            }
        },
        showCasters() {
            nodecg.sendMessage('mainShowCasters');
        },
        async removeCaster(id: string): Promise<void> {
            return nodecg.sendMessage('removeCaster', { id });
        }
    }
});
