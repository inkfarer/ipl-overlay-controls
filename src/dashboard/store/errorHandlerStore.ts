import { App } from 'vue';
import { generateId } from '../../helpers/generateId';
import { defineStore } from 'pinia';

export const useErrorHandlerStore = defineStore('errorHandler', {
    state: () => ({
        recentErrors: {}
    }),
    actions: {
        removeRecentError({ key }: { key: string }): void {
            delete this.recentErrors[key];
        },
        handleError({ err, info }: { err: unknown, info: string }): void {
            console.error(`Got error from '${info}': \n`, err);

            if (Object.keys(this.recentErrors).length >= 2) return;

            const id = generateId();
            this.recentErrors[id] = err;
            window.setTimeout(() => {
                delete this.recentErrors[id];
            }, 25000);
        }
    }
});

export function setUpErrorHandler(app: App<unknown>): void {
    const store = useErrorHandlerStore();
    app.config.errorHandler = (err, vm, info) => {
        store.handleError({ err, info });
    };
}
