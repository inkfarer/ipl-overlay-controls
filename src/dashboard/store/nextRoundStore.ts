import { NextRound } from 'schemas';
import { SetNextRoundRequest } from 'types/messages/rounds';
import { BeginNextMatchRequest } from 'types/messages/activeRound';
import { defineStore } from 'pinia';

const nextRound = nodecg.Replicant<NextRound>('nextRound');

export const nextRoundReps = [ nextRound ];

export interface NextRoundStore {
    nextRound: NextRound;
}

export const useNextRoundStore = defineStore('nextRound', {
    state: () => ({
        nextRound: null
    } as NextRoundStore),
    actions: {
        setShowOnStream(newValue: boolean) {
            nextRound.value.showOnStream = newValue;
        },
        beginNextMatch(request: BeginNextMatchRequest): void {
            nodecg.sendMessage('beginNextMatch', request);
        },
        setNextRound(request: SetNextRoundRequest): void {
            nodecg.sendMessage('setNextRound', request);
        }
    }
});
