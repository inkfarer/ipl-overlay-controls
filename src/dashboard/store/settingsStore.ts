import { LastFmSettings, RadiaSettings, RuntimeConfig } from 'schemas';
import { GameVersion } from 'types/enums/gameVersion';
import { SetGameVersionResponse } from 'types/messages/runtimeConfig';
import { defineStore } from 'pinia';
import { Locale } from 'types/enums/Locale';
import { perGameData } from '../../helpers/gameData/gameData';
import { InterfaceLocale } from 'types/enums/InterfaceLocale';

const lastFmSettings = nodecg.Replicant<LastFmSettings>('lastFmSettings');
const radiaSettings = nodecg.Replicant<RadiaSettings>('radiaSettings');
const runtimeConfig = nodecg.Replicant<RuntimeConfig>('runtimeConfig');

export const settingsReps = [lastFmSettings, radiaSettings, runtimeConfig];

export interface SettingsStore {
    lastFmSettings: LastFmSettings
    radiaSettings: RadiaSettings
    runtimeConfig: RuntimeConfig
}

export const useSettingsStore = defineStore('settings', {
    state: () => ({
        lastFmSettings: {},
        radiaSettings: {
            guildID: null,
            enabled: null,
            updateOnImport: null
        },
        runtimeConfig: null
    } as SettingsStore),
    getters: {
        translatedModeName: state =>
            (mode: string) => (perGameData[state.runtimeConfig.gameVersion]
                .modes[state.runtimeConfig.locale] as Record<string, string>)[mode],
        translatedStageName: state =>
            (stage: string) => (perGameData[state.runtimeConfig.gameVersion]
                .stages[state.runtimeConfig.locale] as Record<string, string>)[stage],
    },
    actions: {
        setLastFmSettings({ newValue }: { newValue: LastFmSettings }): void {
            lastFmSettings.value = newValue;
        },
        setRadiaSettings({ newValue }: { newValue: RadiaSettings }): void {
            radiaSettings.value = newValue;
        },
        setUpdateOnImport(updateOnImport: boolean): void {
            radiaSettings.value.updateOnImport = updateOnImport;
        },
        async attemptRadiaConnection(): Promise<void> {
            return nodecg.sendMessage('retryRadiaAvailabilityCheck');
        },
        setGameVersion(newValue: GameVersion): Promise<SetGameVersionResponse> {
            return nodecg.sendMessage('setGameVersion', { version: newValue });
        },
        setLocale(newValue: Locale): Promise<void> {
            return nodecg.sendMessage('setLocale', newValue);
        },
        setInterfaceLocale(newValue: `${InterfaceLocale}`) {
            runtimeConfig.value.interfaceLocale = newValue;
        }
    }
});
