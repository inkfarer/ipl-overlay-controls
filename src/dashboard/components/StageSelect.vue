<template>
    <ipl-select
        v-model="model"
        :options="options"
    />
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue';
import { useSettingsStore } from '../store/settingsStore';
import { perGameData } from '../../helpers/gameData/gameData';
import { IplSelect } from '@iplsplatoon/vue-components';
import sortBy from 'lodash/sortBy';

export default defineComponent({
    name: 'StageSelect',

    components: { IplSelect },

    props: {
        modelValue: {
            type: String,
            default: null
        }
    },

    emits: ['update:modelValue'],

    setup(props, { emit }) {
        const model = computed({
            get() {
                return props.modelValue;
            },
            set(newValue: string) {
                emit('update:modelValue', newValue);
            }
        });

        const settingsStore = useSettingsStore();

        return {
            model,
            options: computed(() =>
                sortBy(Object.entries(
                    perGameData[settingsStore.runtimeConfig.gameVersion].stages[settingsStore.runtimeConfig.locale])
                    .map(([key, value]) => ({ value: key, name: value })), 'name')),
        };
    }
});
</script>
