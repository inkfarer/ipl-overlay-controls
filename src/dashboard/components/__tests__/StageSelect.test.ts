import { config, shallowMount } from '@vue/test-utils';
import { useSettingsStore } from '../../store/settingsStore';
import { GameVersion } from 'types/enums/gameVersion';
import { Locale } from 'types/enums/Locale';
import { createTestingPinia, TestingPinia } from '@pinia/testing';
import StageSelect from '../StageSelect.vue';
import { IplSelect } from '@iplsplatoon/vue-components';

describe('StageSelect', () => {
    let pinia: TestingPinia;

    beforeEach(() => {
        pinia = createTestingPinia();

        config.global.plugins = [pinia];

        useSettingsStore().runtimeConfig = {
            gameVersion: GameVersion.SPLATOON_2,
            locale: Locale.EN
        };
    });

    it('emits expected event when new option is selected', () => {
        const wrapper = shallowMount(StageSelect);

        wrapper.findComponent<typeof IplSelect>('ipl-select-stub').vm.$emit('update:modelValue', 'New Stage');

        expect(wrapper.emitted()['update:modelValue'][0]).toEqual(['New Stage']);
    });

    it.each(Object.values(Locale))('adds expected options to select when language is %s', locale => {
        useSettingsStore().runtimeConfig.locale = locale;

        const wrapper = shallowMount(StageSelect);

        expect(wrapper.findComponent<typeof IplSelect>('ipl-select-stub').vm.options).toMatchSnapshot();
    });
});
