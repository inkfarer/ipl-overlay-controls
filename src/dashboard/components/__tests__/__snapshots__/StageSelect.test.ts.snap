// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StageSelect adds expected options to select when language is DE 1`] = `
[
  {
    "name": "Abyssal-Museum",
    "value": "Shellendorf Institute",
  },
  {
    "name": "Anchobit Games HQ",
    "value": "Ancho-V Games",
  },
  {
    "name": "Arowana Center",
    "value": "Arowana Mall",
  },
  {
    "name": "Backfisch-Stadion",
    "value": "Goby Arena",
  },
  {
    "name": "Buckelwal-Piste",
    "value": "Humpback Pump Track",
  },
  {
    "name": "Camp Schützenfisch",
    "value": "Camp Triggerfish",
  },
  {
    "name": "Cetacea-Markt",
    "value": "MakoMart",
  },
  {
    "name": "Counterpick",
    "value": "Counterpick",
  },
  {
    "name": "Flunder-Funpark",
    "value": "Wahoo World",
  },
  {
    "name": "Grundel-Pavillon",
    "value": "Skipper Pavilion",
  },
  {
    "name": "Grätenkanal",
    "value": "Snapper Canal",
  },
  {
    "name": "Heilbutt-Hafen",
    "value": "Port Mackerel",
  },
  {
    "name": "Hotel Neothun",
    "value": "New Albacore Hotel",
  },
  {
    "name": "Kofferfisch-Lager",
    "value": "Walleye Warehouse",
  },
  {
    "name": "Korallenviertel",
    "value": "The Reef",
  },
  {
    "name": "Manta Maria",
    "value": "Manta Maria",
  },
  {
    "name": "Molluskelbude",
    "value": "Musselforge Fitness",
  },
  {
    "name": "Muränentürme",
    "value": "Moray Towers",
  },
  {
    "name": "Perlmutt-Akademie",
    "value": "Inkblot Art Academy",
  },
  {
    "name": "Punkasius-Skatepark",
    "value": "Blackbelly Skatepark",
  },
  {
    "name": "Seeigel-Rockbühne",
    "value": "Starfish Mainstage",
  },
  {
    "name": "Steinköhler-Grube",
    "value": "Piranha Pit",
  },
  {
    "name": "Störwerft",
    "value": "Sturgeon Shipyard",
  },
  {
    "name": "Tümmlerkuppel",
    "value": "Kelp Dome",
  },
  {
    "name": "Unbekannte Arena",
    "value": "Unknown Stage",
  },
  {
    "name": "Wandelzone",
    "value": "Shifty Station",
  },
]
`;

exports[`StageSelect adds expected options to select when language is EN 1`] = `
[
  {
    "name": "Ancho-V Games",
    "value": "Ancho-V Games",
  },
  {
    "name": "Arowana Mall",
    "value": "Arowana Mall",
  },
  {
    "name": "Blackbelly Skatepark",
    "value": "Blackbelly Skatepark",
  },
  {
    "name": "Camp Triggerfish",
    "value": "Camp Triggerfish",
  },
  {
    "name": "Counterpick",
    "value": "Counterpick",
  },
  {
    "name": "Goby Arena",
    "value": "Goby Arena",
  },
  {
    "name": "Humpback Pump Track",
    "value": "Humpback Pump Track",
  },
  {
    "name": "Inkblot Art Academy",
    "value": "Inkblot Art Academy",
  },
  {
    "name": "Kelp Dome",
    "value": "Kelp Dome",
  },
  {
    "name": "MakoMart",
    "value": "MakoMart",
  },
  {
    "name": "Manta Maria",
    "value": "Manta Maria",
  },
  {
    "name": "Moray Towers",
    "value": "Moray Towers",
  },
  {
    "name": "Musselforge Fitness",
    "value": "Musselforge Fitness",
  },
  {
    "name": "New Albacore Hotel",
    "value": "New Albacore Hotel",
  },
  {
    "name": "Piranha Pit",
    "value": "Piranha Pit",
  },
  {
    "name": "Port Mackerel",
    "value": "Port Mackerel",
  },
  {
    "name": "Shellendorf Institute",
    "value": "Shellendorf Institute",
  },
  {
    "name": "Shifty Station",
    "value": "Shifty Station",
  },
  {
    "name": "Skipper Pavilion",
    "value": "Skipper Pavilion",
  },
  {
    "name": "Snapper Canal",
    "value": "Snapper Canal",
  },
  {
    "name": "Starfish Mainstage",
    "value": "Starfish Mainstage",
  },
  {
    "name": "Sturgeon Shipyard",
    "value": "Sturgeon Shipyard",
  },
  {
    "name": "The Reef",
    "value": "The Reef",
  },
  {
    "name": "Unknown Stage",
    "value": "Unknown Stage",
  },
  {
    "name": "Wahoo World",
    "value": "Wahoo World",
  },
  {
    "name": "Walleye Warehouse",
    "value": "Walleye Warehouse",
  },
]
`;
